pdns/zonemd.cc

```diff
@@ -65,26 +23,140 @@ void pdns::zonemdVerify(const DNSName& zone, ZoneParserTNG& zpt, bool& validatio
-      soaRecordContent = std::dynamic_pointer_cast<SOARecordContent>(drc);
+    DNSRecord rec;
+    rec.d_name = dnsResourceRecord.qname;
+    rec.d_content = drc;
+    rec.d_type = dnsResourceRecord.qtype;
+    rec.d_class = dnsResourceRecord.qclass;
+    rec.d_ttl = dnsResourceRecord.ttl;
+    rec.d_clen = dnsResourceRecord.content.length(); // XXX is this correct?
+    readRecord(rec);
+  }
+}
+
+void pdns::ZoneMD::readRecords(const vector<DNSRecord>& records)
+{
+  for (auto& record : records) {
+    readRecord(record);
+  }
+}
+
+void pdns::ZoneMD::readRecord(const DNSRecord& record)
+{
+  if (!record.d_name.isPartOf(d_zone) && record.d_name != d_zone) {
+    return;
+  }
+  if (record.d_type == QType::SOA && d_soaRecordContent) {
+    return;
+  }
+
+  if (record.d_name == d_zone) {
+    switch (record.d_type) {
+    case QType::SOA: {
+      d_soaRecordContent = std::dynamic_pointer_cast<SOARecordContent>(record.d_content);
+      if (d_soaRecordContent == nullptr) {
+        throw PDNSException("Invalid SOA record");
+      }
+      break;
+    }
+    case QType::DNSKEY: {
+      auto dnskey = std::dynamic_pointer_cast<DNSKEYRecordContent>(record.d_content);
+      if (dnskey == nullptr) {
+        throw PDNSException("Invalid DNSKEY record");
+      }
+      d_dnskeys.emplace(dnskey);
+      break;
     }
-    if (dnsResourceRecord.qtype == QType::ZONEMD && dnsResourceRecord.qname == zone) {
-      auto zonemd = std::dynamic_pointer_cast<ZONEMDRecordContent>(drc);
-      auto inserted = zonemdRecords.insert({pair(zonemd->d_scheme, zonemd->d_hashalgo), {zonemd, false}});
+    case QType::ZONEMD: {
+      auto zonemd = std::dynamic_pointer_cast<ZONEMDRecordContent>(record.d_content);
+      if (zonemd == nullptr) {
+        throw PDNSException("Invalid ZONEMD record");
+      }
+      auto inserted = d_zonemdRecords.insert({pair(zonemd->d_scheme, zonemd->d_hashalgo), {zonemd, false}});
       if (!inserted.second) {
         // Mark as duplicate
         inserted.first->second.duplicate = true;
       }
+      break;
+    }
+    case QType::RRSIG: {
+      auto rrsig = std::dynamic_pointer_cast<RRSIGRecordContent>(record.d_content);
+      if (rrsig == nullptr) {
+        throw PDNSException("Invalid RRSIG record");
+      }
+      if (rrsig->d_type == QType::NSEC) {
+        d_nsecs.signatures.emplace_back(rrsig);
+      }
+      else if (rrsig->d_type == QType::NSEC3) {
+        d_nsecs3.signatures.emplace_back(rrsig);
+      }
+      d_rrsigs.emplace_back(rrsig);
+      break;
+    }
+    case QType::NSEC: {
+      auto nsec = std::dynamic_pointer_cast<NSECRecordContent>(record.d_content);
+      if (nsec == nullptr) {
+        throw PDNSException("Invalid NSEC record");
+      }
+      d_nsecs.records.emplace(nsec);
+      break;
+    }
+    case QType::NSEC3:
+      // Handled below
+      break;
+    case QType::NSEC3PARAM: {
+      auto param = std::dynamic_pointer_cast<NSEC3PARAMRecordContent>(record.d_content);
```

Comment:
We could also collect the salt and iterations from the first `NSEC3` record we see. In theory we could have a different salt for each record, but I hope no-one will do that for these zones. We could detect it and fall back to collecting every NSEC3, of course.

Reply:
In theory non-IN records can exist and can be handled by ZONEMD, I see now I have to add a few checks and also make sure recToZone handles that correctly (on that case we shold defenitly skip non-IN.

Remains the question if we *want* to allow non-IN records...

Reply:
We'll see how many funyy zones we'll encounter..